src/fifolock.jl

```diff
@@ -0,0 +1,143 @@
+@static if VERSION >= v"1.10"
+
+const LOCKED_BIT = 0b01
+
+"""
+    FIFOLock()
+
+Like Base.ReentrantLock, but with strict FIFO ordering.
+"""
+mutable struct FIFOLock <: AbstractLock
+    @atomic locked_by::Union{Task, Nothing}
+    reentrancy_cnt::UInt32
+    @atomic havelock::UInt8
+    cond_wait::Base.ThreadSynchronizer
+
+    FIFOLock() = new(nothing, 0x0000_0000, 0x00, Base.ThreadSynchronizer())
+end
+
+assert_havelock(l::FIFOLock) = assert_havelock(l, l.locked_by)
+islocked(l::FIFOLock) = (@atomic :monotonic l.havelock) & LOCKED_BIT != 0
+
+# Correctness reasoning:
+# 
+# `havelock` can only be unset with the `cond_wait` lock held.
+# Locking then first tries to set `havelock`; on failure, we
+# acquire the `cond_wait` lock and try to set `havelock` again
+# to ensure that we didn't coincide with an `unlock`. If we
+# fail again, then we are assured that we will not miss an
+# `unlock`, because we hold the `cond_wait` lock. Thus we can
+# safely wait on `cond_wait`.
+#
+# FIFO ordering is ensured in `unlock`, which first acquires
+# the `cond_wait` lock. If `cond_wait`'s wait queue is empty,
+# the lock is released. Otherwise, we pop the first task in
+# the wait queue, transfer ownership to it, schedule it, and
+# return. Thus when one or more tasks are waiting,`havelock`
+# is never reset.
+
+"""
+    trylock(l::FIFOLock)
+
+Try to acquire lock `l`. If successful, return `true`. If the lock is
+held by another task, do not wait and return `false`. 
+"""
+@inline function trylock(l::FIFOLock)
+    ct = current_task()
+    if l.locked_by === ct
```

Comment:
Can you check `===` on an atomic without `@atomic` on this line?
Should it be `@atomic(l.locked_by) === ct`?

Reply:
Awesome, thank you. 👍 (And is the `monotonic (relaxed) ordering` the right ordering for us? I really don't know these details. IMO every access to the atomics could use a comment explaining why that ordering is the right choice

Reply:
It can be very difficult to reason about, and really requires some understanding of relaxed memory ordering _as implemented_ on some architectures like Power. This is one reason why I often ask Jameson and Valentin to review such work, and I would love Andre's review here as well.

In general, relaxed ordering means that there will be a total order on all atomic operations to this atomic variable, but loads/stores to/from other variables can be reordered around this atomic operation. This is fine for this case.